File: LeetCode-Java/Unique Binary Search Trees.java
public class Solution {
    public int numTrees(int n) {
        int[] result = new int[n + 1];
        for(int i = 0;i <= n;i ++)
        {
            if(i == 0 || i == 1) result[i] = 1;
            else
            {
                result[i] = 0;
                for(int j = 1; j <= i;j ++)
                {
                    result[i] += result[j - 1] * result[i - j];
                }
            }
        }
        return result[n];
    }
}

File: Leetcode-cpp/UniqueBinarySearchTrees.cpp
class Solution {
public:
    int numTrees(int n) {
        vector<int> result;
        result.push_back(1);
        for(int i = 1;i <= n;i ++)
        {
            int cResult = 0;
            for(int j = 1;j <= i;j ++)
            {
                cResult += result[j - 1] * result[i - j ];
            }
            result.push_back(cResult);
        }
        return result[n];
    }
};
